File: src/lexer/lexer.ts
import { Token, TokenType } from "./token"

const keywords: Map<string, TokenType> = new Map([
    ["import", TokenType.Import],
    ["true", TokenType.True],
    ["false", TokenType.False],
    ["nil", TokenType.Nil],
    ["var", TokenType.Var],
    ["const", TokenType.Const],
    ["func", TokenType.Func],
    ["if", TokenType.If],
    ["elif", TokenType.Elif],
    ["else", TokenType.Else],
    ["for", TokenType.For],
    ["while", TokenType.While],
    ["break", TokenType.Break],
    ["continue", TokenType.Continue],
    ["return", TokenType.Return],
    ["struct", TokenType.Struct],
    ["public", TokenType.Public],
    ["private", TokenType.Private],
    ["int", TokenType.Int],
    ["float", TokenType.Float],
    ["string", TokenType.String],
    ["bool", TokenType.Bool],
    ["void", TokenType.Void],
    ["char", TokenType.Char],
])

const operators: Map<string, TokenType> = new Map([
    ["+", TokenType.Add],
    ["-", TokenType.Minus],
    ["*", TokenType.Mul],
    ["/", TokenType.Div],
    ["%", TokenType.Mod],
    ["=", TokenType.Assign],
    ["==", TokenType.Equal],
    ["!=", TokenType.NotEqual],
    [">", TokenType.GreaterThan],
    [">=", TokenType.GreaterEqual],
    ["<", TokenType.LessThan],
    ["<=", TokenType.LessEqual],
    ["&&", TokenType.And],
    ["||", TokenType.Or],
    ["!", TokenType.Not],
    ["&", TokenType.BAnd],
    ["|", TokenType.BOr],
    ["^", TokenType.BXor],
    ["<<", TokenType.LMove],
    [">>", TokenType.RMove],
    ["++", TokenType.Increment],
    ["--", TokenType.Decrement],
    [",", TokenType.Comma],
    [";", TokenType.Semicolon],
    [":", TokenType.Colon],
    ["{", TokenType.LBrace],
    ["}", TokenType.RBrace],
    ["[", TokenType.LBracket],
    ["]", TokenType.RBracket],
    ["(", TokenType.LParen],
    [")", TokenType.RParen],
    [".", TokenType.Dot],
])

export class Lexer {
    private file: string
    private src: string
    private line: number
    private column: number
    private pos: number

    constructor(src: string, file: string) {
        this.file = file
        this.src = src
        this.line = 1
        this.column = 1
        this.pos = 0
    }

    nextToken(): Token {
        this.skipWhitespace()
        if (this.pos >= this.src.length) {
            return this.newToken(TokenType.EndOfFile)
        }
        if (operators.has(this.src[this.pos])) {
            if (operators.has(this.src.substring(this.pos, this.pos + 2))) {
                const op = this.next(2)
                return this.newToken(operators.get(op)!)
            } else {
                const op = this.next()
                return this.newToken(operators.get(op)!)
            }
        }
        if (isLetter(this.src[this.pos])) {
            return this.scanIdentifier()
        }
        if (isDigit(this.src[this.pos])) {
            return this.scanNumber()
        }
        if (this.src[this.pos] == '"') {
            return this.scanString()
        }
        if (this.src[this.pos] == "'") {
            return this.scanChar()
        }
        return this.newToken(TokenType.Unknown)
    }

    private scanString(): Token {
        let result = ""
        this.next()
        while (this.src[this.pos] != '"') {
            if (this.src[this.pos] == "\\") {
                this.next()
                if (this.src[this.pos] == "n") {
                    result += "\n"
                } else if (this.src[this.pos] == "t") {
                    result += "\t"
                } else if (this.src[this.pos] == "r") {
                    result += "\r"
                } else if (this.src[this.pos] == "b") {
                    result += "\b"
                } else if (this.src[this.pos] == "f") {
                    result += "\f"
                } else if (this.src[this.pos] == '"') {
                    result += '"'
                } else if (this.src[this.pos] == "\\") {
                    result += "\\"
                } else {
                    result += this.src[this.pos]
                }
            } else {
                result += this.next()
            }
        }
        this.next()
        return this.newToken(TokenType.String, result)
    }

    private scanChar(): Token {
        this.next()
        let ch = this.next()
        if (ch == "\\") {
            ch = this.next()
            if (ch == "n") {
                ch = "\n"
            } else if (ch == "t") {
                ch = "\t"
            } else if (ch == "r") {
                ch = "\r"
            } else if (ch == "b") {
                ch = "\b"
            } else if (ch == "f") {
                ch = "\f"
            } else if (ch == "'") {
                ch = "'"
            } else if (ch == "\\") {
                ch = "\\"
            } else {
                ch = this.src[this.pos]
            }
        }
        if (this.src[this.pos] != "'") {
            return this.newToken(TokenType.Unknown)
        }
        this.next()
        return this.newToken(TokenType.Char, ch)
    }

    private scanNumber(): Token {
        let result = ""
        let dotCount = 0
        while (isDigit(this.src[this.pos]) || this.src[this.pos] == ".") {
            if (this.src[this.pos] == ".") {
                dotCount += 1
                if (dotCount > 1) {
                    break
                }
            }
            result += this.next()
        }
        if (dotCount == 0) {
            return this.newToken(TokenType.Int, result)
        } else {
            return this.newToken(TokenType.Float, result)
        }
    }

    private scanIdentifier(): Token {
        let result = ""
        while (isLetter(this.src[this.pos])) {
            result += this.next()
        }

        if (keywords.has(result)) {
            return this.newToken(keywords.get(result)!)
        } else {
            return this.newToken(TokenType.Identifier, result)
        }
    }

    private skipWhitespace(): void {
        while (
            isWhitespace(this.src[this.pos]) ||
            this.test("//") ||
            this.test("/*")
        ) {
            if (this.src[this.pos] == "\n") {
                this.line += 1
                this.column = 1
            } else if (this.test("//") || this.test("/*")) {
                this.skipComment()
            } else {
                this.column += 1
            }
            this.pos += 1
        }
    }

    private skipComment(): void {
        if (this.test("/*")) {
            this.pos += 2
            while (!this.test("*/")) {
                if (this.src[this.pos] == "\n") {
                    this.line += 1
                    this.column = 1
                } else {
                    this.column += 1
                }
                this.pos += 1
            }
            this.pos += 2
        } else if (this.test("//")) {
            while (this.src[this.pos] != "\n") {
                this.pos += 1
            }
            this.line += 1
            this.column = 1
            this.pos += 1
        }
    }

    private next(n: number = 1): string {
        const result = this.src.slice(this.pos, this.pos + n)
        this.pos += n
        this.column += n
        return result
    }

    private test(ch: string): boolean {
        return this.src.slice(this.pos, this.pos + ch.length) == ch
    }

    private newToken(type: TokenType, literal: string = ""): Token {
        return {
            type: type,
            literal: literal,
            line: this.line,
            column: this.column,
            file: this.file,
        }
    }
}

function isDigit(ch: string): boolean {
    return /[0-9]/.test(ch)
}

function isLetter(ch: string): boolean {
    return /[a-zA-Z_]/.test(ch)
}

function isWhitespace(ch: string): boolean {
    return /\s/.test(ch)
}

File: src/compiler/parser/Lexer.kt
package compiler.parser

class Lexer(private val src: String, private val file: String) {
    private var pos: Int = 0
    private var ln: Int = 1
    private var col: Int = 1

    fun nextToken(): Token {
        skipWhitespace()
        if (pos >= src.length) {
            return Token(TokenType.EndOfFile, "", ln, col, file)
        }
        if (operators.containsKey(src.substring(pos, pos + 1))) {
            if (operators.containsKey(src.substring(pos, pos + 2))) {
                val op = next(2)
                val tokenType = operators[op] ?: TokenType.Unknown
                return Token(tokenType, op, ln, col, file)
            } else {
                val op = next(1)
                val tokenType = operators[op] ?: TokenType.Unknown
                return Token(tokenType, op, ln, col, file)
            }
        }
        if (isLetter(src[pos])) {
            return scanIdentifier()
        }
        if (isDigit(src[pos])) {
            return scanNumber()
        }
        if (src[pos] == '"') {
            return scanString()
        }
        if (src[pos] == '\'') {
            return scanChar()
        }
        return Token(TokenType.Unknown, "", ln, col, file)
    }

    private fun scanChar(): Token {
        next()
        var ch = next()
        if (ch == "\\") {
            ch = next()
            if (ch == "n") {
                ch = "\n"
            } else if (ch == "t") {
                ch = "\t"
            } else if (ch == "r") {
                ch = "\r"
            } else if (ch == "b") {
                ch = "\b"
            } else if (ch == "'") {
                ch = "'"
            } else if (ch == "\\") {
                ch = "\\"
            } else {
                ch = src.substring(pos, pos + 1)
            }
        }
        if (src[pos] != '\'') {
            return Token(TokenType.Unknown, "", ln, col, file)
        }
        next()
        return Token(TokenType.CharLiteral, ch, ln, col, file)
    }

    private fun scanString(): Token {
        var result = ""
        next()
        while (src[pos] != '"') {
            if (src[pos] == '\\') {
                next()
                if (src[pos] == 'n') {
                    result += "\n"
                } else if (src[pos] == 't') {
                    result += "\t"
                } else if (src[pos] == 'r') {
                    result += "\r"
                } else if (src[pos] == 'b') {
                    result += "\b"
                } else if (src[pos] == '"') {
                    result += '"'
                } else if (src[pos] == '\\') {
                    result += "\\"
                } else {
                    result += src[pos]
                }
            } else {
                result += next()
            }
        }
        next()
        return Token(TokenType.StringLiteral, result, ln, col, file)
    }

    private fun scanNumber(): Token {
        var res = ""
        var dotNum = 0
        while (isDigit(src[pos]) || src[pos] == '.') {
            if (src[pos] == '.') {
                dotNum++
                if (dotNum > 1) {
                    break
                }
            }
            res += next()
        }
        if (dotNum == 0) {
            return Token(TokenType.IntLiteral, res, ln, col, file)
        }
        return Token(TokenType.FloatLiteral, res, ln, col, file)
    }

    private fun scanIdentifier(): Token {
        var res = next()
        while (isLetterOrDigit(src[pos]) || src[pos] == '_') {
            res += next()
        }
        return Token(keywords[res] ?: TokenType.Identifier, res, ln, col, file)
    }

    private fun next(n: Int = 1): String {
        val res = src.substring(pos, pos + n)
        pos += n
        col++
        return res
    }

    private fun skipWhitespace() {
        while (isWhitespace(src[pos]) || test("//") || test("/*")) {
            if (src[pos] == '\n') {
                ln++
                col = 1
                pos++
            } else if (test("/*") || test("//")) {
                skipComment()
            } else {
                col++
                pos++
            }
        }
    }

    private fun skipComment() {
        if (test("/*")) {
            pos += 2
            while (!test("*/")) {
                if (src[pos] == '\n') {
                    ln++
                    col = 1
                }
                pos++
            }
            pos += 2
        } else if (test("//")) {
            while (src[pos] != '\n') {
                pos++
            }
            ln++
            col = 1
            pos++
        }
    }

    private fun test(st: String): Boolean {
        return src.startsWith(st, pos)
    }
}

private fun isWhitespace(ch: Char): Boolean {
    return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r'
}

private fun isLetter(ch: Char): Boolean {
    return ch.isLetter() || ch == '_'
}

private fun isLetterOrDigit(ch: Char): Boolean {
    return ch.isLetterOrDigit() || ch == '_'
}

private fun isDigit(ch: Char): Boolean {
    return ch.isDigit()
}

private val keywords = mapOf<String, TokenType>(
    "import" to TokenType.Import,
    "as" to TokenType.As,
    "true" to TokenType.True,
    "false" to TokenType.False,
    "null" to TokenType.Null,
    "var" to TokenType.Var,
    "const" to TokenType.Const,
    "func" to TokenType.Func,
    "if" to TokenType.If,
    "elif" to TokenType.Elif,
    "else" to TokenType.Else,
    "for" to TokenType.For,
    "while" to TokenType.While,
    "break" to TokenType.Break,
    "continue" to TokenType.Continue,
    "return" to TokenType.Return,
    "struct" to TokenType.Struct,
    "public" to TokenType.Public,
    "private" to TokenType.Private,
    "int" to TokenType.Int,
    "float" to TokenType.Float,
    "string" to TokenType.String,
    "bool" to TokenType.Bool,
    "char" to TokenType.Char
)

private val operators = mapOf<String, TokenType>(
    "(" to TokenType.LParen,
    ")" to TokenType.RParen,
    "{" to TokenType.LBrace,
    "}" to TokenType.RBrace,
    "[" to TokenType.LBracket,
    "]" to TokenType.RBracket,
    "." to TokenType.Dot,
    "," to TokenType.Comma,
    ":" to TokenType.Colon,
    "" to TokenType.Semicolon,
    "=" to TokenType.Assign,
    "+" to TokenType.Add,
    "-" to TokenType.Minus,
    "*" to TokenType.Mul,
    "/" to TokenType.Div,
    "%" to TokenType.Mod,
    "<" to TokenType.LessThan,
    "<=" to TokenType.LessEqual,
    ">" to TokenType.GreaterThan,
    ">=" to TokenType.GreaterEqual,
    "==" to TokenType.Equal,
    "!=" to TokenType.NotEqual,
    "&&" to TokenType.And,
    "||" to TokenType.Or,
    "!" to TokenType.Not,
    "&" to TokenType.BAnd,
    "|" to TokenType.BOr,
    "^" to TokenType.BXor,
    "~" to TokenType.BNot,
    "<<" to TokenType.LMove,
    ">>" to TokenType.RMove,
    "++" to TokenType.Increment,
    "--" to TokenType.Decrement
)
